### DISCUSSION TITLE
React Ref on Deck.gl (not exposing all props)

### DISCUSSION THREAD

**Discussion by JavLG** [visgl/deck.gl#6646] [Q&A]
Hi,
I'm trying to access the setProps method via ref on React Deck.gl implementation. When i create a ref based on the Deck component, i only get the following object:

{pickObject: f(), pickMultipleObjects: f(), pickObjects: f()}

Using version:
  "deck.gl": "^8.6.3"

Has anyone encountered this issue? I'm willing to update my layers by setting the corresponding props to shown layers but cant find a way to do it. I tried updateTriggers grained approach, but still cant figure out how to get these to update layers getMethods.

I'd appreciate all the help you guys can provide. Even a Github repo where there's a similar implementation can be really helpful.

Thanks and KR.
```

`
// Inside component

...
console.log(deckRef) // returns {current:{pickObject: f(), pickMultipleObjects: f(), pickObjects: f()}}

return(
 <DeckGL
        ref={deckRef}
        initialViewState={viewport}
        onViewStateChange={(e) => {
          setZoomFactor(e.viewState.zoom);
          setMapScreenSize(e.viewState);
        }}
        controller={true}
        layers={cartoLayers}
        ContextProvider={_MapContext.Provider}
      >
...
<MapComp/>
</DeckGL>
`

```

**Comment by Pessimistress** [comment]
Use the React `layers` prop to update layers. Anything you set by calling the underlying `setProps` will be overwritten when your react component rerenders.

**Reply by JavLG** [reply]
Hi @Pessimistress, first of all, thanks for the feedback. I've tried to use setProps method on layers and I get the following error:
<img width="1032" alt="image" src="https://user-images.githubusercontent.com/33915099/154174335-fb1243d6-20ce-41e2-a2ca-5577294e5a30.png">

Are layers supposed to provide the setProps method after i instanciate them as in:

const layer = new CartoLayer(....)


Would I be able to provide props to this layer? Pardon my inexperience in the field. I'm not sure if you meant to have state manage the layer props and just pass new layers with new props whenever i need some updates to be done.

Regards.

**Reply by Pessimistress** [reply]
You create new layer instances with the desired props on each rerender, just like you do with other React components. Please read https://deck.gl/docs/developer-guide/using-layers.

**Reply by JavLG** [reply]
If i provide the connection query to data, I cant figure out how to extend the data object as shown in the documentation. My layer looks like this. 

```
  const Tileset = new CartoLayer({
    id: `${tableName}-tilesetlayer`,
    type: MAP_TYPES.TILESET,
    connection: 'test-production',
    data: `source.PROJECT_TILESETS.${tableName}`,
    getFillColor: colorContinuous({
      attr: 'NDVI',
      domain: defaultVectorDomain,
      colors: getColorbarColorsRGB(defaultVectorDomain),
    }),
    updateTriggers: {
      getFillColor: d => d.zoomFactor >= 14 ? [[111, 111, 111, 111]] : [[255, 255, 255, 255]],
    },
    pickable: true,
    minZoom: 15,
  });
```

I'm still figuring out how to pass to `data` the connection query and my state based const `zoomFactor`. I've seen the docs and there isn't any piece stating how to extend the data object. This is the key point i need to understand, thank for all the help already though.



Thanks in advance.

**Reply by Pessimistress** [reply]
I don't know what "the connection query" is and that might be a question for @padawannn 

You are using `updateTriggers` wrong. It should look like this:

```js
updateTriggers: {
  getFillColor: [zoomFactor]
}
```

See https://deck.gl/docs/api-reference/core/layer#updatetriggers

**Reply by felixpalmer** [reply]
@JavLG it would be useful if you could provide a full example, e.g. as a jsfiddle. Also I don't understand what you mean by "extend the data object as shown in the documentation", could you link to the relevant docs you are referencing?

**Reply by padawannn** [reply]
@JavLG I don't know what you mean with `the connection query`. Maybe are you talking about the parameters to establish the connection with your database? If this is so, the CartoLayer doesn't work like this, you need to register the connection in the [CARTO workspace](https://carto.com/carto3/). In any case, as @felixpalmer said an example would help.

**Reply by JavLG** [reply]
I was able to update layers by correctly setting the update triggers as @Pessimistress suggested. I will build the sandbox today or tomorrow and upload it here to show the implementation so you gyus can close the thread. Thanks for all the support guys.

**Comment by JavLG** [comment]
This is the actual implementation. The function LayerManager is being called on my Map Component, returning and setting to a state the actual layers, which the map renders. Zoom factor should update what gradient or colors are shown but i still cant figure out if this usage is correct.

```
import chroma from 'chroma-js';
import { MAP_TYPES, colorContinuous, colorBins, CartoLayer } from '@deck.gl/carto';
import { times } from 'lodash';

function LayerManager(tableName, zoomFactor) {
  // Select shown layers with properties
  const PolygonLayer = new CartoLayer({
    id: `${tableName}-polygonlayer`,
    type: MAP_TYPES.TABLE,
    connection: 'test-production',
    data: `source.PROJECTS.${tableName}`,
    geoColumn: 'cropGeometry',
    getFillColor:
      zoomFactor < 14
        ? colorBins({
          attr: 'vegetationPercent',
          domain: defaultVectorDomain,
          colors: getColorbarColorsRGB(defaultVectorDomain, 'alt'),
        })
        : colorContinuous({
            attr: 'vegetationPercent',
            domain: defaultVectorDomain,
            colors: getColorbarColorsRGB(defaultVectorDomain),
          }),
    updateTriggers: {
      getFillColor: [zoomFactor],
    },
    pickable: true,
  });

  const Tileset = new CartoLayer({
    id: `${tableName}-tilesetlayer`,
    type: MAP_TYPES.TILESET,
    connection: 'test-production',
    data: `source.PROJECT_TILESETS.${bqTableName}`,
    getFillColor: colorContinuous({
      attr: 'NDVI',
      domain: defaultVectorDomain,
      colors: getColorbarColorsRGB(defaultVectorDomain),
    }),
    getFillOpacity: zoomFactor > 16 ? 0.5 : 0.1, // Is this working on CartoLayers? I've seen on GeoViz that fillOpacity can be used.
    getStrokeOpacity: zoomFactor > 16 ? 0 : 1, // Is this working on CartoLayers? I've seen on GeoViz that strokeOpacity can be used.
    updateTriggers: {
      getFillColor: [zoomFactor],
      getFillOpacity: [zoomFactor],
      getStrokeOpacity: [zoomFactor],
    },
  });

  return [Tileset, PolygonLayer];
}

export default LayerManager;

// COLORS AND FILL LOGIC
const scale = chroma.scale(['#d73027', '#eb9514', '#ffe795', '#8CCE67', '#1a9850']);
const scale2 = chroma.scale(['#f9ddda', '#f2b9c4', '#e597b9', '#ce78b3', '#ad5fad']);
const defaultVectorDomain = [0, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1];

const getColorbarColorsRGB = (domain, type = 'normal') => {
  const [min, max] = [domain[0], domain[10]];

  const stepSize = (max - min) / 9;
  const stops = times(10, (idx) => min + stepSize * idx);

  scale.domain([min, max]);
  return type === 'normal'
    ? stops.map((stop) => scale(stop).rgb())
    : stops.map((stop) => scale2(stop).rgb());
};

```
Any ideas why i'm not being able to determine the colour of layers directly to getFillColor as [[255,255,255,255]]  or getting them to update correctly?

**Reply by felixpalmer** [reply]
`getFillOpacity` and `getStrokeOpacity` are not valid accessors. You should pass the alpha value as the fourth component of the `getFillColor` accessor. `CartoLayer` supports all the props you will find in https://deck.gl/docs/api-reference/layers/geojson-layer - take a look there to see what you need to pass

**Reply by JavLG** [reply]
Thank you, I´ll use the alpha to either hide or show color. Is there a way to either show or hide each layer based on update triggers dependencies? I noticed that passing new layers without the ones i'm willing to remove depending on a state update won't take them from the screen... @felixpalmer

**Reply by Pessimistress** [reply]
Use the `visible` prop.

**Comment by felixpalmer** [comment]
If you update the layers array with a new set of layers only those will be rendered. If that is not happening, you most likely have a bug somewhere in your code.

Separately, another approach you could take when it comes to toggling layer visibility is https://deck.gl/docs/api-reference/core/deck#layerfilter
